### PR TITLE
Feature/ees 820 amendments allow add footnotes

### PR DESCRIPTION
**Pull request by shakes90**
This PR covers not just the functionality to allow the addition, deletion & amendment of footnotes for a release amendment but also the function to add & delete (not amend) datafiles for a release amendment.

### CODE REVIEW

File: src/explore-education-statistics-frontend/src/modules/methodologies/MethodologyPage.tsx [hunk 1]
```diff
@@ -131,7 +131,8 @@ const MethodologyPage: NextPage<Props> = ({ data }) => {
                   caption={caption}
                   key={order}
                 >
-                  <SectionBlocks content={content} />
+                  2
```
Comment (ntsim) [inline src/explore-education-statistics-frontend/src/modules/methodologies/MethodologyPage.tsx:134]: This looks like a typo

File: src/explore-education-statistics-admin/src/components/editable/EditableBlockRenderer.tsx [hunk 2]
```diff
@@ -6,6 +6,7 @@ import DataBlockRenderer from '@common/modules/find-statistics/components/DataBl
 import React, { useCallback, useMemo } from 'react';
 
 interface Props {
+  releaseId: string;
```
Comment (ntsim) [inline src/explore-education-statistics-admin/src/components/editable/EditableBlockRenderer.tsx:9]: Would make this optional

File: src/explore-education-statistics-admin/src/components/editable/EditableBlockRenderer.tsx [hunk 3]
```diff
@@ -41,6 +43,7 @@ function EditableBlockRenderer({
         <div className="dfe-content-overflow">
           <EditableBlockWrapper onDelete={editable ? handleDelete : undefined}>
             <DataBlockRenderer
+              releaseId={releaseId}
```
Comment (ntsim) [inline src/explore-education-statistics-admin/src/components/editable/EditableBlockRenderer.tsx:46]: Would return null on this branch of the switch statement if there isn't a `releaseId`. We shouldn't be trying to pass through an empty string.

File: src/explore-education-statistics-frontend/src/modules/methodologies/MethodologyPage.tsx [hunk 4]
```diff
@@ -110,7 +110,7 @@ const MethodologyPage: NextPage<Props> = ({ data }) => {
                   </div>
 
                   <div className="govuk-grid-column-three-quarters">
-                    <SectionBlocks content={content} />
+                    <SectionBlocks releaseId="" content={content} />
```
Comment (ntsim) [inline src/explore-education-statistics-frontend/src/modules/methodologies/MethodologyPage.tsx:113]: Shouldn't be passing through empty strings. Make `releaseId` optional in `SectionBlocks`.

File: src/GovUk.Education.ExploreEducationStatistics.Admin/Services/FootnoteService.cs [hunk 5]
```diff
@@ -484,5 +391,27 @@ private static IQueryable<Footnote> HydrateFootnote(IQueryable<Footnote> query)
                 .Include(footnote => footnote.Subjects)
                 .SingleOrDefault();
         }
+        
+        private async Task<bool> IsFootnoteExclusiveToReleaseAsync(Guid releaseId, Guid footnoteId)
```
[the remark this answers is not in the document]
Comment (shakes90) [reply src/GovUk.Education.ExploreEducationStatistics.Admin/Services/FootnoteService.cs:395]: From an api perspective you could potentially pass an id that was for another release and delete the footnote incorrectly so I think I'll leave as is.

File: src/GovUk.Education.ExploreEducationStatistics.Data.Api/Controllers/FastTrackController.cs [hunk 6]
```diff
@@ -28,5 +28,16 @@ public async Task<ActionResult<FastTrackViewModel>> GetAsync(string id)
 
             return NotFound();
         }
+
+        [HttpGet("release/{releaseId}/{id}")]
+        public async Task<ActionResult<FastTrackViewModel>> GetAsync(Guid releaseId, string id)
```
[the remark this answers is not in the document]
Comment (shakes90) [reply src/GovUk.Education.ExploreEducationStatistics.Data.Api/Controllers/FastTrackController.cs:33]: agreed - cheers

File: src/GovUk.Education.ExploreEducationStatistics.Data.Model/Migrations/20200528153827_EES-820_Delete_ReleaseFootnotes.cs [hunk 7]
```diff
@@ -0,0 +1,23 @@
...
+        private const string PreviousVersionMigrationId = "20200528124909";
+
+        protected override void Up(MigrationBuilder migrationBuilder)
+        {
+            migrationBuilder.SqlFromFile(MigrationsPath, $"{MigrationId}_Routine_DropAndCreateRelease.sql");
+        }
+
+        protected override void Down(MigrationBuilder migrationBuilder)
+        {
+            // Revert to the version in the previous migration 20200303144410_Routine_DropAndCreateRelease.sql
```
[the remark this answers is not in the document]
Comment (shakes90) [reply src/GovUk.Education.ExploreEducationStatistics.Data.Model/Migrations/20200528153827_EES-820_Delete_ReleaseFootnotes.cs:19]: corrected

File: src/GovUk.Education.ExploreEducationStatistics.Admin.Tests/Controllers/Api/Statistics/TableBuilderControllerTests.cs [hunk 8]
```diff
@@ -5,6 +5,7 @@
 using GovUk.Education.ExploreEducationStatistics.Common.Model.Data.Query;
 using GovUk.Education.ExploreEducationStatistics.Data.Services.Interfaces;
 using GovUk.Education.ExploreEducationStatistics.Data.Services.ViewModels;
+using Microsoft.EntityFrameworkCore.Storage.ValueConversion;
```
[the remark this answers is not in the document]
Comment (shakes90) [reply src/GovUk.Education.ExploreEducationStatistics.Admin.Tests/Controllers/Api/Statistics/TableBuilderControllerTests.cs:8]: No I didn't - cheers

File: src/GovUk.Education.ExploreEducationStatistics.Admin.Tests/Controllers/Api/Statistics/TableBuilderControllerTests.cs [hunk 9]
```diff
@@ -14,6 +15,7 @@ namespace GovUk.Education.ExploreEducationStatistics.Admin.Tests.Controllers.Api
     public class TableBuilderControllerTests
     {
         private readonly TableBuilderController _controller;
+        private readonly Guid releaseId = new Guid("03730cff-22d5-446c-8971-68921e933b50");
```
[the remark this answers is not in the document]
Comment (shakes90) [reply src/GovUk.Education.ExploreEducationStatistics.Admin.Tests/Controllers/Api/Statistics/TableBuilderControllerTests.cs:18]: Changed

File: src/GovUk.Education.ExploreEducationStatistics.Admin/Services/FootnoteService.cs [hunk 10]
```diff
@@ -86,96 +82,110 @@ public class FootnoteService : AbstractRepository<Footnote, Guid>, IFootnoteServ
...
+                    
+                    if (await IsFootnoteExclusiveToReleaseAsync(releaseId, footnote.Id))
+                    {
+                        DeleteEntities(footnote.Subjects);
+                        DeleteEntities(footnote.Filters);
+                        DeleteEntities(footnote.FilterGroups);
+                        DeleteEntities(footnote.FilterItems);
+                        DeleteEntities(footnote.Indicators);
+
+                        Remove(id);
```
[the remark this answers is not in the document]
Comment (shakes90) [reply src/GovUk.Education.ExploreEducationStatistics.Admin/Services/FootnoteService.cs:115]: yep -done

File: src/GovUk.Education.ExploreEducationStatistics.Admin/Services/FootnoteService.cs [hunk 11]
```diff
@@ -484,5 +391,27 @@ private static IQueryable<Footnote> HydrateFootnote(IQueryable<Footnote> query)
...
+        private async Task<bool> IsFootnoteExclusiveToReleaseAsync(Guid releaseId, Guid footnoteId)
+        {
+            var otherFootnoteReferences = await _context
+                .ReleaseFootnote
+                .CountAsync(rf => rf.FootnoteId == footnoteId && rf.ReleaseId != releaseId);
+
+            return otherFootnoteReferences == 0;
+        }
+        
+        private async Task<bool> DeleteReleaseFootnoteLinkAsync(Guid releaseId, Guid footnoteId)
```
[the remark this answers is not in the document]
Comment (shakes90) [reply src/GovUk.Education.ExploreEducationStatistics.Admin/Services/FootnoteService.cs]: done

File: src/GovUk.Education.ExploreEducationStatistics.Data.Model/Migrations/20200526100824_EES-820_Add_ReleasesFootnote_ForeignKeys.cs [hunk 12]
```diff
@@ -0,0 +1,17 @@
+﻿using Microsoft.EntityFrameworkCore.Migrations;
+
+namespace GovUk.Education.ExploreEducationStatistics.Data.Model.Migrations
+{
+    public partial class EES820_Add_ReleasesFootnote_ForeignKeys : Migration
+    {
+        protected override void Up(MigrationBuilder migrationBuilder)
+        {
+
```
[the remark this answers is not in the document]
Comment (shakes90) [reply src/GovUk.Education.ExploreEducationStatistics.Data.Model/Migrations/20200526100824_EES-820_Add_ReleasesFootnote_ForeignKeys.cs]: yep it's redundant - removed - thanks

**Pull request closed without merging**